pkg/graveler/graveler.go

```diff
@@ -1745,6 +1779,38 @@ func (g *KVGraveler) Commit(ctx context.Context, repositoryID RepositoryID, bran
 	return newCommitID, nil
 }
 
+func (g *KVGraveler) retryBranchUpdate(ctx context.Context, repositoryID RepositoryID, branchID BranchID, f BranchUpdateFunc) error {
+	const (
+		maxInterval = 5
+		setTries    = 3
+	)
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = maxInterval * time.Second
+
+	var try int
+	err := backoff.Retry(func() error {
+		// TODO(3586) - if the branch commit id hasn't changed, update the fields instead of fail
+		err := g.RefManager.BranchUpdate(ctx, repositoryID, branchID, f)
+		if err != nil {
+			if !errors.Is(err, kv.ErrPredicateFailed) {
+				return backoff.Permanent(err)
+			}
+			if try < setTries-1 {
+				g.log.WithField("try", try+1).
+					WithField("branchID", branchID).
+					Info("Retrying update branch")
+				try += 1
```

Comment:
By moving this one to the start of the func we can prevent all the -1 +1 while using try

Reply:
we want to lon only from the second time when we retry